components/datetime/src/provider/neo.rs

```diff
@@ -0,0 +1,172 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::pattern::{self, runtime};
+use icu_provider::prelude::*;
+use tinystr::UnvalidatedTinyAsciiStr;
+use zerovec::{VarZeroVec, ZeroMap};
+
+/// Symbols used for representing the year name
+///
+/// This uses an auxiliary key for length.
+///
+/// <div class="stab unstable">
+/// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
+/// including in SemVer minor releases. While the serde representation of data structs is guaranteed
+/// to be stable, their Rust representation might not be. Use with caution.
+/// </div>
+#[icu_provider::data_struct(marker(YearSymbolsV1Marker, "datetime/symbols/years@1"))]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+pub enum YearSymbolsV1<'data> {
+    /// This calendar uses eras with numeric years, this stores the era names mapped from
+    /// era code to the name
+    Eras(
+        #[cfg_attr(feature = "serde", serde(borrow))]
+        ZeroMap<'data, UnvalidatedTinyAsciiStr<16>, str>,
```

Comment:
Thought: Wish we could use shorter than 16. Variable length like you had before might be fine. We can benchmark later.